dali/operators/reader/loader/nemo_asr_loader.cc

```diff
@@ -0,0 +1,212 @@
+  const auto str_end = str.find_last_not_of(whitespace);
+  const auto str_len = str_end - str_begin + 1;
+  return str.substr(str_begin, str_len);
+}
+
+std::string NormalizeText(std::string& text) {
+  // Remove trailing and leading whitespace
+  auto norm_text = trim(text);
+
+  // Convert to lowercase
+  for (auto &c : norm_text) {
+    c = std::tolower(c);
+  }
+  return norm_text;
+}
+
+void ParseManifest(std::vector<NemoAsrEntry> &entries, std::istream& manifest_file,
+                   float min_duration, float max_duration, bool normalize_text) {
+  std::string line;
+  while (std::getline(manifest_file, line)) {
+    detail::LookaheadParser parser(const_cast<char*>(line.c_str()));
+    if (parser.PeekType() != kObjectType) {
+      DALI_WARN(make_string("Skipping invalid manifest line: ", line));
+      continue;
+    }
+    parser.EnterObject();
+    NemoAsrEntry entry;
+    while (const char* key = parser.NextObjectKey()) {
+      if (0 == detail::safe_strcmp(key, "audio_filepath")) {
+        entry.audio_filepath = parser.GetString();
+      } else if (0 == detail::safe_strcmp(key, "duration")) {
+        entry.duration = parser.GetDouble();
+      } else if (0 == detail::safe_strcmp(key, "offset")) {
+        entry.offset = parser.GetDouble();
+        DALI_WARN("Handing of ``offset`` is not yet implemented and will be ignored.");
+      } else if (0 == detail::safe_strcmp(key, "text")) {
+        entry.text = parser.GetString();
+        if (normalize_text)
+          entry.text = NormalizeText(entry.text);
+      } else {
+        parser.SkipValue();
+      }
+    }
+    if (entry.audio_filepath.empty()) {
+      DALI_WARN(make_string("Skipping manifest line without an audio filepath: ", line));
+      continue;
+    }
+
+    if ((max_duration > 0.0f && entry.duration > max_duration) ||
+        (min_duration > 0.0f && entry.duration < min_duration)) {
+      continue;  // skipping sample
+    }
+
+    entries.emplace_back(std::move(entry));
+  }
+}
+
+}  // namespace detail
+
+void NemoAsrLoader::PrepareMetadataImpl() {
+  for (auto &manifest_filepath : manifest_filepaths_) {
+    std::ifstream fstream(manifest_filepath);
+    DALI_ENFORCE(fstream,
+                 make_string("Could not open NEMO ASR manifest file: \"", manifest_filepath, "\""));
+    detail::ParseManifest(entries_, fstream, max_duration_, normalize_text_);
+  }
+
+  DALI_ENFORCE(Size() > 0, "No files found.");
+  if (shuffle_) {
+    // seeded with hardcoded value to get
+    // the same sequence on every shard
+    std::mt19937 g(kDaliDataloaderSeed);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+  Reset(true);
+}
+
+void NemoAsrLoader::Reset(bool wrap_to_shard) {
+  current_index_ = wrap_to_shard ? start_index(shard_id_, num_shards_, Size()) : 0;
+  current_epoch_++;
+
+  if (shuffle_after_epoch_) {
+    std::mt19937 g(kDaliDataloaderSeed + current_epoch_);
+    std::shuffle(entries_.begin(), entries_.end(), g);
```

Comment:
Is it possible that there are outstanding reads when this is called? You're passing `entry` to a lambda by reference, so this it will get corrupted if another thread reaches this point.

Reply:
If you are asking about `Reset` then this part of the code is single threaded. No concurrency can/should happen.

Reply:
Unless I missed something it's like this:
It's called from MoveToNextShard and that, in turn, is called from inside ReadSample. ReadSample has a lambda captures a _reference_ to a particular index in `entries_` - that lambda is executed later and will see stale data in its `entry`.